needle.pm

```diff
@@ -276,13 +276,34 @@ sub has_tag {
 }
 
 sub has_property {
-    my ($self, $tag) = @_;
-    for my $t (@{$self->{properties}}) {
-        return 1 if ($t eq $tag);
+    my ($self, $property_name) = @_;
+    for my $property (@{$self->{properties}}) {
+        if (ref($property) eq "HASH") {
+            return 1 if ($property->{name} eq $property_name);
+        }
+        else {
+            return 1 if ($property eq $property_name);
+        }
```

Comment:
```perl
return 1 if ($property->{name} // $property) eq $property_name);
```

should work or even to replace the whole for-loop and return block as well:

```perl
return grep { ($_->{name} // $_) eq $property_name } @{$self->{properties}};

Reply:
If I modify the code as your suggestion, there will be a warning message "Cannot use a string as HASH...". 
Current code maybe looks a bit stupid, but I think there is no warning message and simpler to read. Can I keep them?

Reply:
IMHO the explicit for loop should be avoided, hence I suggested grep. You can put the check for hash within the grep if it helps

Reply:
Why should the explicit for loop be avoided?  I do not know how is `grep` implemented in Perl.  If there is also a for in its implement, I prefer to keep the `for` loop.  In my opinion, it is more readable.

Reply:
To be honest: The `grep` line looks harder to read than the more verbose but simpler for-loop. Especially for those who are not that familiar with Perl. So from my point of view: Keep it simple. (Actually I don't care much but I wouldn't require you to change anything here.)

Reply:
I don't think we can get by learning nice perl without using `grep` or `map` where it fits but I would trust more experienced persons here of course:

@kraih what's your take on this? Explicit for-loop or grep? :)

Reply:
Yes, understanding`map` and `grep` is considered an important milestone in learning Perl. I would generally use a `grep` one-liner over a bigger for loop here.

Reply:
I used `grep` instead of the `for` loop.